packages/parcel-runtime/src/index.ts

```diff
@@ -1,39 +1,127 @@
 import { Runtime } from "@parcel/plugin"
 import fs from "fs"
-import path from "path"
+import path, { basename, dirname, join } from "path"
 
-import type { HmrData } from "./types"
+import { vLog } from "@plasmo/utils/logging"
 
-const hmrRuntimeCode = fs.readFileSync(
-  path.join(__dirname, "./hmr-runtime.js"),
-  "utf8"
+import type { RuntimeData } from "./types"
+
+const [pageRuntime, scriptRuntime, backgroundServiceRuntime] = [
+  "page-runtime",
+  "script-runtime",
+  "background-service-runtime"
+].map((runtimeName) =>
+  fs.readFileSync(path.join(__dirname, `./${runtimeName}.js`), "utf8")
 )
 
 export default new Runtime({
-  apply({ bundle, options }) {
+  async loadConfig({ config }) {
+    const pkg =
+      (await config.getPackage()) ||
+      (await config
+        .getConfigFrom<{
+          dependencies: Record<string, string>
+          devDependencies: Record<string, string>
+        }>(
+          join(process.env.PLASMO_SRC_DIR, "lab"), // parcel only look up
+          ["package.json"],
+          {
+            exclude: true
+          }
+        )
+        .then((cfg) => cfg?.contents))
+
+    const hasReact = !!pkg?.dependencies?.react || !!pkg?.devDependencies?.react
+
+    return {
+      hasReact
+    }
+  },
+
+  apply({ bundle, options, config, bundleGraph }) {
+    if (bundle.name === "manifest.json") {
+      const asset = bundle.getMainEntry()
+      if (asset?.meta.webextEntry !== true) {
+        return
+      }
+
+      // Hack to bust packager cache when any descendants update
+      const descendants = []
+      bundleGraph.traverseBundles((b) => {
+        descendants.push(b.id)
+      }, bundle)
+
+      return {
+        filePath: __filename,
+        code: JSON.stringify(descendants),
+        isEntry: true
+      }
+    }
+
     if (
       bundle.type !== "js" ||
       !options.hmrOptions ||
       bundle.env.isLibrary ||
       bundle.env.isWorklet() ||
+      options.mode !== "development" ||
       bundle.env.sourceType === "script"
     ) {
       return
     }
 
-    const hmrData: HmrData = {
-      host: options.hmrOptions.host,
-      port: options.hmrOptions.port,
-      secure: !!(options.serveOptions && options.serveOptions.https),
+    const entryFilePath = bundle.getMainEntry().filePath
+
+    const isPlasmo = entryFilePath.includes(".plasmo")
+
+    const isBackground =
+      entryFilePath.startsWith(
+        join(process.env.PLASMO_SRC_DIR, "background")
+      ) || entryFilePath.endsWith("plasmo-default-background.ts")
+
+    const isPlasmoSrc =
+      isPlasmo ||
+      isBackground ||
+      entryFilePath.startsWith(join(process.env.PLASMO_SRC_DIR, "content"))
```

Comment:
This will fail when someone calls their script content_blahblah (or anything else that starts with content). I think it'd be better if this were more precise.

Reply:
I was gonna leave a comment there to explain why that's not needed - this runtime only processes entry files that were added by the transformer. These files basically need to be vetted by our pre-compiler which generates the manifest, thus this check is mainly a filter against parcel's other runtime files (react-refresh wrapper etc...)